db/migrate/20210212223928_handle_nil_variants.rb

```diff
@@ -0,0 +1,63 @@
+class HandleNilVariants < ActiveRecord::Migration
+   def up
+     Spree::Variant.where(product_id: nil).destroy_all
+     Spree::Variant.where(unit_value: [nil, Float::NAN]).find_each do |variant|
+       if can_update_with_callbacks?(variant)
+         variant.unit_value = 1
+         begin
+           variant.save
+         rescue
+           variant.update_column(:unit_value, 1)
+         end
+       else
+         variant.update_column(:unit_value, 1)
+       end
+     end
+     Spree::Variant.where(weight: Float::NAN).find_each do |variant|
+       if can_update_with_callbacks?(variant)
+         variant.weight = 0
+         begin
+           variant.save
+         rescue
+           variant.update_column(:weight, 0)
+         end
+       else
+         variant.update_column(:weight, 0)
+       end
+     end
+     change_column_null :spree_variants, :unit_value, false, 1
+     change_column_default :spree_variants, :unit_value, 1
+     change_column_default :spree_variants, :weight, 0.0
+
+     if !constraint_present?("check_unit_value_for_nan")
+       execute "ALTER TABLE spree_variants ADD CONSTRAINT check_unit_value_for_nan CHECK (unit_value <> 'NaN')"
```

Comment:
This is a really good point. Once we're on Rails 6.1 we can [add them](https://github.com/rails/rails/pull/31323) but we can't make sure that all instances have them until then. 

The alternative you're suggesting would be a validation on the model, right? Ruby's BigDecimal class has a `#nan?` method.

Reply:
Something like this, I think: https://github.com/openfoodfoundation/openfoodnetwork/pull/6926

Reply:
> This will apply to current production servers but it's not persisted in the schema and won't be there in new instances.

good catch @mkllnk . We'll need to switch to `db/structure.sql` for that, which I'm totally in favor of.